### PR TITLE
Add batchprocessor support for client metadata

### PR DESCRIPTION
**Pull request by jmacd**
**Description:** 

Re-do of https://github.com/open-telemetry/opentelemetry-collector/pull/7325

**Link to tracking Issue:** #4544 

**Testing:** One new test.

**Documentation:** README updated.

### CODE REVIEW

**Comment by jmacd** [comment]
I screwed up the review history of https://github.com/open-telemetry/opentelemetry-collector/pull/7325#issuecomment-1523920075, so have created this replacement PR. Sorry I force-pushed and lost the review-comments by mistake.

**Review by codeboten — COMMENTED** [review]
Pinging approvers of the original PR to review here @jpkrohling @astencel-sumo @ericmustin

File: processor/batchprocessor/factory.go [hunk 1]
```diff
@@ -30,6 +30,11 @@ const (
 
 	defaultSendBatchSize = uint32(8192)
 	defaultTimeout       = 200 * time.Millisecond
+
+	// defaultMetadataCardinalityLimit should be set to the number
+	// of metadata configurations the user expects to submit to
+	// the collector.
+	defaultMetadataCardinalityLimit = 1000
```
Comment (codeboten) [inline processor/batchprocessor/factory.go:37]: There was a discussion in the SIG call yesterday about the defaults for the batch processor being too high, i wonder if we should start with a lower limit than this and increase it if users find it to be too low.

Comment (jmacd) [reply processor/batchprocessor/factory.go:37]: I raised it from 100 to 1000 based on the comment here: https://github.com/open-telemetry/opentelemetry-collector/pull/7325#discussion_r1161766124

**Review by jpkrohling — APPROVED** [review]

**Review by gramidt — APPROVED** [review]

**Review by codeboten — APPROVED** [review]

**Review by dmitryax — APPROVED** [review]

**Pull request merged**